pkg/controllers/sandbox/unsandboxsubcluster_reconciler.go

```diff
@@ -0,0 +1,267 @@
+		err := r.updateSandboxStatus(ctx, sb, scs)
+		if err != nil {
+			r.Log.Error(err, "failed to update sandbox status", "sandbox", sb, "new subclusters", scs)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// reconcileSandboxConfigMap will update/delete sandbox config map if it expires, this function will return
+// an error and a boolean to indicate if sandbox config map is deleted
+func (r *UnsandboxSubclusterReconciler) reconcileSandboxConfigMap(ctx context.Context) (error, bool) {
+	sbName := r.ConfigMap.Data[vapi.SandboxNameKey]
+	cmName := r.ConfigMap.Name
+	sb := r.Vdb.GetSandboxStatus(sbName)
+	// if the sandbox doesn't have any subclusters, we delete the config map
+	if sb == nil || len(sb.Subclusters) == 0 {
+		err := r.Client.Delete(ctx, r.ConfigMap)
+		if err != nil {
+			r.Log.Error(err, "failed to delete expired sandbox config map", "configMapName", cmName)
+			return err, false
+		}
+		r.Log.Info("deleted expired sandbox config map", "configMapName", cmName)
+		return nil, true
+	}
+	if r.ConfigMap.Annotations[vmeta.SandboxControllerUnsandboxTriggerID] != "" {
+		unsandboxSbScMap := r.Vdb.GenSandboxSubclusterMapForUnsandbox()
+		_, found := unsandboxSbScMap[sbName]
+		// if the subclusters in the sandbox does not need to be unsandboxed, we remove
+		// unsandbox trigger ID from the config map
+		if !found {
+			delete(r.ConfigMap.Annotations, vmeta.SandboxControllerUnsandboxTriggerID)
+			err := r.Client.Update(ctx, r.ConfigMap)
+			if err != nil {
+				r.Log.Error(err, "failed to remove unsandbox trigger ID from sandbox config map", "configMapName", cmName)
+				return err, false
+			}
+			r.Log.Info("Successfully removed unsandbox trigger ID from sandbox config map", "configMapName", cmName)
+		}
+	}
+
+	return nil, false
+}
+
+// unsandboxSubclusters will unsandbox subclusters inside the sandbox from sandbox config map
+func (r *UnsandboxSubclusterReconciler) unsandboxSubclusters(ctx context.Context) (ctrl.Result, error) {
+	// find an initiator to call vclusterOps
+	initiatorIP, ok := r.PFacts.FindFirstPrimaryUpPodIP()
+	if ok {
+		r.InitiatorIP = initiatorIP
+	} else {
+		r.Log.Info("Requeue because there are no UP nodes in main cluster to execute unsandbox operation")
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	err := r.executeUnsandboxCommand(ctx)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	r.PFacts.Invalidate()
+
+	return ctrl.Result{}, nil
+}
+
+// executeUnsandboxCommand will move subclusters from a sandbox to main cluster, update sandbox
+// status, and delete the config map
+func (r *UnsandboxSubclusterReconciler) executeUnsandboxCommand(ctx context.Context) error {
+	unsandboxSbScMap := r.Vdb.GenSandboxSubclusterMapForUnsandbox()
+	sbName := r.ConfigMap.Data[vapi.SandboxNameKey]
+	scs, found := unsandboxSbScMap[sbName]
+	if !found {
+		r.Log.Info("Ignore the config map because the sandbox inside it does not need to be unsandboxed")
+		return nil
+	}
+	succeedScs := []string{}
+	for _, sc := range scs {
```

Comment:
Does the order of the subclusters matter? I would think that we need to unsandbox the one denoted as the primary last.

Reply:
The order doesn't matter. The unsandbox is done through main cluster nodes. If we unsandbox the primary subcluster first, the secondary subcluster can still be alive and unsandboxed. I tested this case using vcluster cli.

Reply:
I guess they go into read-only mode then.